File: src/org/jetbrains/jet/j2k/ast/Type.java
package org.jetbrains.jet.j2k.ast;

import org.jetbrains.annotations.NotNull;

/**
 * @author ignatov
 */
public abstract class Type extends Element {
  public static final Type EMPTY_TYPE = new EmptyType();
  private boolean myNullable = true;

  @NotNull
  @Override
  public Kind getKind() {
    return Kind.TYPE;
  }

  public void setNullable(boolean nullable) {
    myNullable = nullable;
  }

  public boolean isNullable() {
    return myNullable;
  }

  String isNullableStr() {
    return isNullable() ? QUEST : EMPTY;
  }

  /**
   * @author ignatov
   */
  private static class EmptyType extends Type {
    @NotNull
    @Override
    public String toKotlin() {
      return "UNRESOLVED_TYPE";
    }

    @Override
    public boolean isNullable() {
      return false;
    }
  }
}

File: src/org/jetbrains/jet/j2k/ast/types/Type.kt
package org.jetbrains.jet.j2k.ast.types

import org.jetbrains.jet.j2k.ast.Element
import org.jetbrains.jet.j2k.ast.INode

public abstract class Type(val nullable: Boolean) : Element() {
    public open fun convertedToNotNull() : Type {
        if (nullable) throw UnsupportedOperationException("convertedToNotNull must be defined")
        return this
    }

    public open fun isNullableStr() : String? {
        return (if (nullable)
            "?"
        else
            "")
    }
}
